packages/cli/generators/datasource/templates/datasource.ts.ejs

```diff
@@ -0,0 +1,8 @@
+import {inject} from '@loopback/core';
+import {juggler, DataSource} from '@loopback/repository';
+
+export class <%= className %>DataSource extends juggler.DataSource {
+  constructor(@inject('datasources.config.<%= name %>') dsConfig: DataSource) {
```

Comment:
`dsConfig: DataSource` looks suspicious to me.

```
export interface DataSource {
  name: string; // Name of the data source
  connector?: Connector; // The underlying connector instance

  settings: AnyObject; // Settings
  // tslint:disable-next-line:no-any
  [property: string]: any; // Other properties that vary by connectors
}
```

1) Configuration wise, we allow developers to specify the connector as a string, e.g. `'mysql'` instead of `require('loopback-connector-mysql')`.

2) DataSource configuration does not contain any `settings` object, these settings (e.g. `url`, `database`, etc.) are top-level properties of the config object.

I think you should use `juggler.Options` type instead, see [juggler.DataSource constructors](https://github.com/strongloop/loopback-datasource-juggler/blob/77f11cda3b6e9e39c1cc039808180129a23d67ff/types/datasource.d.ts#L87-L89) and [definition of Options](https://github.com/strongloop/loopback-datasource-juggler/blob/77f11cda3b6e9e39c1cc039808180129a23d67ff/types/common.d.ts#L13-L16.

Reply:
To make these datasource classes easier to use in tests, I think `dsConfig` should have a default value.

```
import * as CONFIG from './<%= name %>.json';

export {CONFIG};

export class <%= className %>DataSource extends juggler.DataSource {
  constructor(@inject('datasources.config.<%= name %>') dsConfig: juggler.Options = CONFIG) {
    super(dsConfig);
  }
}
```

Thoughts?

Reply:
Hmm ... I had used DI for the config based on your comment in the boot spike. That said looking at the case of overriding the dsConfig by binding a partial value (such as setting `file` to `undefined` in `examples/todo`) I think it might be better to merge the default config from the file with the config that is bound to Context ... only question is that is that what a user would expect?

```
import * as CONFIG from './<%= name %>.json';

export class <%= className %>DataSource extends juggler.DataSource {
  constructor(@inject('datasources.config.<%= name %>') dsConfig: juggler.Options = {}) {
    super(Object.assign(CONFIG, dsConfig));
  }
}
```

Reply:
^ With that my only concern is what a developer expects ... do they expect config that lives in context to be used or is a merge what they expect?

Reply:
With regards to the first comment, I wasn't too sure of the purpose of the `settings` object since I did realize that the settings are actually a top level object. I can change it to `juggler.Options` -- 

What is the purpose of the `DataSource` interface as it's defined here? -- Is it just a remnant of something before the refactor / move to `juggler.DataSource`?

Reply:
> What is the purpose of the DataSource interface as it's defined here? -- Is it just a remnant of something before the refactor / move to juggler.DataSource?

No idea. @raymondfeng  could you please clarify?

> I had used DI for the config based on your comment in the boot spike.

Sure, I am fine with using DI, I think the Todo example app is already showing why we need that.

> I think it might be better to merge the default config from the file with the config that is bound to Context .

That's a bad idea based on my past experience with loopback-boot. The trouble with merging is that users wishing to obtain a specific datasource configuration must unset all unwanted properties provided in the "default" config, which is pretty much impossible.

For example, consider the case where you want to configure a MySQL datasource to use a specific user+password+database combination while ensuring that no other datasource options are set. Now imagine somebody modifies the JSON configuration and adds let's say `charset` property. Now we need to update the code wanting only user+password+database code to exclude `charset` from the original config.

Thoughts?

Reply:
I would like to use `import` for the json directly within the generated TS file (supported by 2.9.x - https://blogs.msdn.microsoft.com/typescript/2018/05/31/announcing-typescript-2-9/#json-imports)

**my-db.datasource.ts**
```ts
import config from './my-db.json'; // Unfortunately it cannot be `my-db.datasource.json`.

export class MyDBDataSource extends juggler.DataSource {
  constructor() {
    super(config);
  }
}
```